### PR TITLE
Document Crawlspaces

### PR DESCRIPTION
**Pull request by engineer124**
A followup to BgCamInfo. Crawlspaces are primarily covered by 4 systems:
- camera
- player
- bgcheck
- onepointdemo

This documents crawlspaces across these different systems. Two comments to add on:
1. Camera is cursed with reused temps for different variables. It was decided a while ago to try and use macros to deal with this in `Camera_Subj4` but I feel I've grown less fond of this approach over time. I've left it in for now, but open to suggestions/ideas
2. The data for crawlspaces in `camera_data.c` is very strange and contains what appears to be unused data. This is likely due to devs using it for one implementation, but dropping it for another. It's hard to say for certain. See this PR for more info: https://github.com/zeldaret/oot/pull/1225

### CODE REVIEW

**Review by Dragorn421 — COMMENTED** [review]
Nice, finally in PR :partying_face: 

I want to name bgcheck surfacetype-like things/flags, so that will conflict with your `bgcheck_wall_` additions

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 1]
```diff
@@ -228,8 +228,8 @@ void func_8084BDFC(Player* this, PlayState* play);
 void func_8084BF1C(Player* this, PlayState* play);
 void Player_UpdateCommon(Player* this, PlayState* play, Input* input);
 void func_8084C5F8(Player* this, PlayState* play);
-void func_8084C760(Player* this, PlayState* play);
-void func_8084C81C(Player* this, PlayState* play);
+void Player_UpdateInsideCrawlspace(Player* this, PlayState* play);
+void Player_UpdateLeavingCrawlspace(Player* this, PlayState* play);
```
Comment (fig02) [inline src/overlays/actors/ovl_player_actor/z_player.c:232]: these are 674 "action funcs", so not sure I am on board with naming it "UpdateX"
a comment above these functions might be better until docs pick up and we have a naming scheme

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 2]
```diff
@@ -6342,84 +6343,96 @@ s32 func_8083EC18(Player* this, PlayState* play, u32 arg2) {
...
+        // Perpindicular XZ-Distance from player pos to crawlspace line
+        // Uses y-component of crossproduct formula for the distance from a point to a line
+        xVertex2 = ((this->actor.world.pos.x - xVertex1) * COLPOLY_GET_NORMAL(wallPoly->normal.z)) -
+                   ((this->actor.world.pos.z - zVertex1) * COLPOLY_GET_NORMAL(wallPoly->normal.x));
 
-        if (fabsf(phi_f2) < 8.0f) {
-            this->stateFlags2 |= PLAYER_STATE2_16;
+        if (fabsf(xVertex2) < 8.0f) {
+            // Give do-action prompt to "Enter on A" for the crawlspace
+            this->stateFlags2 |= PLAYER_STATE2_CAN_ENTER_CRAWLSPACE;
```
Comment (fig02) [inline src/overlays/actors/ovl_player_actor/z_player.c:6407]: since this state flag is *only* controlling the do action label on the A button, I feel like this name might make it seem like it has more responsibility than it does

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 3]
```diff
@@ -6468,25 +6481,37 @@ s32 func_8083F524(PlayState* play, Player* this) {
     return func_8083F360(play, this, 26.0f, this->ageProperties->unk_38 + 5.0f, 30.0f, 0.0f);
 }
 
-s32 func_8083F570(Player* this, PlayState* play) {
-    s16 temp;
+/**
+ * Two exit walls are placed at each end of the crawlspace, separate to the two entrance walls used to enter the
+ * crawlspace. These front and back exit walls are slightly more interior in the crawlspace relative to the front and
```
Comment (fig02) [inline src/overlays/actors/ovl_player_actor/z_player.c:6486]: "slightly more interior" doesnt read right to me, though maybe its fine? Id say slightly more inside

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 4]
```diff
@@ -6439,13 +6452,13 @@ s32 func_8083F360(PlayState* play, Player* this, f32 arg1, f32 arg2, f32 arg3, f
...
-                                &sp78)) {
+                                &wallBgId)) {
         wallPoly = this->actor.wallPoly;
 
         this->actor.bgCheckFlags |= BGCHECKFLAG_PLAYER_WALL_INTERACT;
-        this->actor.wallBgId = sp78;
+        this->actor.wallBgId = wallBgId;
 
-        D_808535F0 = func_80041DB8(&play->colCtx, wallPoly, sp78);
+        sWallFlags = SurfaceType_GetWallFlags(&play->colCtx, wallPoly, wallBgId);
```
Comment (fig02) [inline src/overlays/actors/ovl_player_actor/z_player.c:6461]: possibly `sInteractWallFlags`? makes it more clear its the flags of the wall youre interacting with. and is also set with `BGCHECKFLAG_PLAYER_WALL_INTERACT`

File: src/overlays/actors/ovl_player_actor/z_player.c (hunk 3 above)
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_player_actor/z_player.c:6486]: Or just "further into".

File: src/overlays/actors/ovl_player_actor/z_player.c (hunk 1 above)
Comment (Dragorn421) [reply src/overlays/actors/ovl_player_actor/z_player.c:232]: I'm all for a `Player_Action674_` prefix

File: src/code/z_camera.c [hunk 5]
```diff
@@ -4351,100 +4352,131 @@ s32 Camera_Subj4(Camera* camera) {
...
+    lineOffset = Math_CosS(rwData->lineOffsetAngle);
+
+    // VEC3F_LERPIMPDST(eye, eye, &sp98, fabsf(temp_f16))
+    eye->x += (vEyeTarget.x - eye->x) * fabsf(lineOffset);
+    eye->y += (vEyeTarget.y - eye->y) * fabsf(lineOffset);
+    eye->z += (vEyeTarget.z - eye->z) * fabsf(lineOffset);
+
+    // When camera reaches the peak of offset and starts to move down
+    // && alternating cycles (sfx plays only every 2nd cycle)
+    if ((rwData->lineOffsetPrev < lineOffset) && !rwData->isSfxOff) {
```
Comment (Dragorn421) [inline src/code/z_camera.c:4452]: Doesn't `lineOffsetPrev < lineOffset` check for lineOffset starting to go up (at its lowest) instead of going down (at a peak)?

I'm not sure if `atEyeTargetOffset.pitch = 0x238C` is upwards or downwards otherwise I could tell what I'd expect from the vertical camera movement

File: src/overlays/actors/ovl_player_actor/z_player.c (hunk 1 above)
Comment (engineer124) [reply src/overlays/actors/ovl_player_actor/z_player.c]: I've left these unnamed for now, the comments above should be enough until this is worked out

**Review by fig02 — APPROVED** [review]

File: src/code/z_camera.c [hunk 6]
```diff
@@ -4341,18 +4345,17 @@ s32 Camera_Subj4(Camera* camera) {
...
-    s16 temp_a0;
-    f32 tx;
+    Vec3f temp1;
+    Vec3f zoomAtTarget;
+    f32 temp2;
     Player* player;
-    PosRot* playerPosRot = &camera->playerPosRot;
+    f32 eyeLerp;
+    PosRot playerPosRot;
+    VecSph atEyeTargetOffset;
```
Comment (Dragorn421) [inline src/code/z_camera.c:4354]: I'm not sure `atEyeTargetOffset` is accurate?

File: src/code/z_camera.c (hunk 5 above)
Comment (Dragorn421) [reply src/code/z_camera.c]: so `atEyeTargetOffset.pitch = 0x238C` is a VecSphGeo so points upwards
so vEyeTarget is above eyeNext
so with eyeLerp=0 eye=eyeNext is lower than vEyeTarget
and with eyeLerp=+-1 eye=vEyeTarget is higher than eyeNext

`if ((eyeLerp > rwData->eyeLerp) && !rwData->isSfxOff) {`
passes when eyeLerp starts raising, i.e. raising from -1
i.e. indeed at a peak height of the eye pos
the "alternating" (every 2 up/down) is because eye=vEyeTarget at eyeLerp=1 too

TLDR it's okay

File: src/code/z_camera.c (hunk 6 above)
Comment (Dragorn421) [reply src/code/z_camera.c:4354]: Oh so `atEyeTargetOffset` is the contraction of `atTargetOffset` and `eyeTargetOffset`

I was reading it like "target offset from at to eye"

Could it be just "targetOffset"? Or at least comment on the name, I think it's unclear enough (I would fall into the trap again)

File: src/code/z_camera.c [hunk 7]
```diff
@@ -4367,101 +4370,136 @@ s32 Camera_Subj4(Camera* camera) {
...
-    at->x = eye->x + (Math_SinS(temp_a0) * 10.0f);
+    camera->player->actor.shape.rot.y = targetOffset.yaw;
+
+    eyeLerp = ((240.0f * eyeLerp) * (rwData->xzSpeed * 0.416667f));
+    xzOffsetTimer = eyeLerp + rwData->yaw;
+
+    at->x = eye->x + (Math_SinS(xzOffsetTimer) * 10.0f);
     at->y = eye->y;
-    at->z = eye->z + (Math_CosS(temp_a0) * 10.0f);
+    at->z = eye->z + (Math_CosS(xzOffsetTimer) * 10.0f);
```
Comment (Dragorn421) [inline src/code/z_camera.c:4493]: `xzOffsetTimer` is the yaw from eye to at

typically `xzOffsetTimer ~ rwData->yaw` which is the crawlspace orientation (forwards or backwards I'm not sure, btw `Camera_XZAngle` may have "from" and "to" swapped... I'd say forwards but not sure)

as for the 10.0f and whatever goes into `eyeLerp`, to me it's just values that Work Well :tm:

Comment (Dragorn421) [reply src/code/z_camera.c:4493]: btw `rwData->yaw` could be nice as `rwData->yawForwards` or `rwData->yawBackwards` depending on which it is?

Comment (engineer124) [reply src/code/z_camera.c]: Nice, yeah that's clearer. I double checked and `rwData->yaw` is always facing towards the forward crawling direction, so I've also renamed it and added a comment

File: include/z64camera.h [hunk 8]
```diff
@@ -894,7 +894,7 @@ typedef struct {
     /* 0x28 */ f32 eyeLerp;
     /* 0x2C */ s16 eyeLerpPhase;
     /* 0x2E */ s16 isSfxOff;
-    /* 0x30 */ s16 yaw;
+    /* 0x30 */ s16 forwardsYaw; // yaw to the forwards crawling direction
```
Comment (fig02) [inline include/z64camera.h:897]: "forwardYaw" sounds more natural to me. looking it up it seems that the differences between the two are pretty muddy and can be used interchangeably, so its not a big deal wither way

Comment (engineer124) [reply include/z64camera.h:897]: I didn't even notice that I also did the reverse order of what was suggested, `yawForward` vs. `forwardYaw` (vs.  `yawForwards` vs. `forwardsYaw`), I'm going to take a hunch that it's not consistent

Comment (fig02) [reply include/z64camera.h:897]: forwardYaw sounds even better to me personally, but yeah whatever works

**Review by fig02 — APPROVED** [review]

**Review by Dragorn421 — APPROVED** [review]

File: include/z64player.h [hunk 9]
```diff
@@ -432,9 +432,9 @@ typedef struct {
 #define PLAYER_STATE2_13 (1 << 13)
 #define PLAYER_STATE2_14 (1 << 14)
 #define PLAYER_STATE2_15 (1 << 15)
-#define PLAYER_STATE2_16 (1 << 16)
+#define PLAYER_STATE2_DO_ACTION_ENTER (1 << 16) // Turns on the "Enter On A" DoAction
 #define PLAYER_STATE2_17 (1 << 17)
-#define PLAYER_STATE2_18 (1 << 18)
+#define PLAYER_STATE2_CRAWLING (1 << 18) // Crawling though a crawlspace
```
Comment (Dragorn421) [inline include/z64player.h:437]:
```suggestion
#define PLAYER_STATE2_CRAWLING (1 << 18) // Crawling through a crawlspace
```
:pensive:

**Review by Dragorn421 — APPROVED** [review]

File: src/code/z_camera.c [hunk 10]
```diff
@@ -4353,6 +4353,9 @@ s32 Camera_Subj3(Camera* camera) {
     return 1;
 }
 
+#define vCrawlSpaceBackPos temp1
+#define vPlayerDistToFront temp2
```
Comment (hensldm) [inline src/code/z_camera.c:4357]: Feel like these would be better inside the function, would flow better since they get swapped out half-way through.

File: src/code/z_camera.c [hunk 11]
```diff
@@ -4398,101 +4400,136 @@ s32 Camera_Subj4(Camera* camera) {
...
+    at->z = eye->z + (Math_CosS(eyeToAtYaw) * 10.0f);
+
     camera->roll = Camera_LERPCeilS(0, camera->roll, 0.5f, 0xA);
-    return 1;
+
+    return true;
 }
 
+#undef vEyeTarget
+#undef vZoomTimer
```
Comment (hensldm) [inline src/code/z_camera.c:4531]: Same here, feel it would be better in the function. Commenting which temp they correspond to would also be nice, like you did for `vCrawlSpaceBackPos` and `vPlayerDistToFront` when you swapped the defines.

File: src/code/z_camera.c [hunk 12]
```diff
@@ -4398,101 +4400,136 @@ s32 Camera_Subj4(Camera* camera) {
...
+
     camera->player->actor.world.pos = *eyeNext;
     camera->player->actor.world.pos.y = camera->playerGroundY;
-    camera->player->actor.shape.rot.y = sp64.yaw;
-    temp_f16 = ((240.0f * temp_f16) * (rwData->unk_24 * 0.416667f));
-    temp_a0 = temp_f16 + rwData->unk_30;
-    at->x = eye->x + (Math_SinS(temp_a0) * 10.0f);
+    camera->player->actor.shape.rot.y = targetOffset.yaw;
+
+    eyeLerp = ((240.0f * eyeLerp) * (rwData->xzSpeed * 0.416667f));
```
Comment (hensldm) [inline src/code/z_camera.c:4518]:
```suggestion
    eyeLerp = (240.0f * eyeLerp) * (rwData->xzSpeed * 0.416667f);
```
Small

File: src/code/z_camera.c (hunk 10 above)
Comment (hensldm) [reply src/code/z_camera.c:4357]: Also, out of curiosity, what is meant by the `v` prefix?

**Review by hensldm — COMMENTED** [review]
A couple thoughts, mainly around the temp defines.

File: src/code/z_camera.c (hunk 10 above)
Comment (fig02) [reply src/code/z_camera.c:4357]: its used in a few places in the codebase to indicate a variable alias

Comment (hensldm) [reply src/code/z_camera.c:4357]: Interesting, looks like the other places don't bother undefing them, just have multiple defines resolve to the same thing. Not sure if this should conform to that style?

Comment (fig02) [reply src/code/z_camera.c:4357]: yea i dont think they particularly need to be undefined. the v prefix is intended to make them unique anyway, and if its changed to be scoped in the function there wont be any clashes either.
at the very least like you mentioned, if we do decide to start undefining them, its something we'd have to do in multiple places

Comment (engineer124) [reply src/code/z_camera.c]: Do `#define` preprocessor commands follow scope? So if I define something inside the function, will it undefine internally once it's out of scope? I suppose I can just test it myself too

Comment (Dragorn421) [reply src/code/z_camera.c]: No defines don't follow C scopes, the preprocessor is unaware of C syntax

**Review by hensldm — APPROVED** [review]

**Review by Dragorn421 — APPROVED** [review]
@fig02 we mergin' ?

**Review by EllipticEllipsis — COMMENTED** [review]
Can't say I love the macros but I suppose it's clearer than trying to give the temps combined names.

**Comment by fig02** [comment]
@EllipticEllipsis if your review has been addressed to your liking we should be good to merge

**Review by EllipticEllipsis — APPROVED** [review]
